### PR TITLE
Parse dot-notation validation rules into nested array shapes

### LINKED ISSUE #624: Parse dot-notation validation rules for nested array shape inference
### Problem:
When using Laravel's dot notation for nested arrays or wildcards (e.g., author.name or comments.*.id), Psalm loses the context of the deeply nested structure. It typically falls back to a generic array<string, mixed> or fails to map the nested shapes correctly.

```
public function store(Request $request)
{
    $data = $request->validate([
        'title' => 'required|string',
        'author.name' => 'required|string',
        'author.email' => 'required|email',
        'tags.*.id' => 'required|integer',
    ]);

    // Psalm currently infers something like: array<string, mixed> if author and tags are specified explicitly as 'array'
    // Or it flattens the keys literally instead of nesting them if not.
    
    // We want Psalm to infer:
    // array{
    //     title: string, 
    //     author: array{name: string, email: string}, 
    //     tags: list<array{id: int}>
    // }
}
```

### Proposed solution
Introduce a parser to convert dot-notation strings into a structured tree format. By evaluating these keys as a tree rather than flat strings, the plugin can dynamically construct the correct, nested Psalm array shapes (including handling * wildcards as lists/iterables).

I've prepared a PR for this feature: #625

### PR DESCRIPTION
**Pull request by alies-dev**
## Issue to Solve

When Laravel validation uses dot-notation rules (e.g., `'address.city' => 'required|string'`), the plugin produced flat keys like `array{'address.city': string}` instead of the nested shape `array{address: array{city: string}}` that `validated()` actually returns.

## Related

Closes #624
Supersedes #625

## Solution Description

Refactored `ValidatedTypeHandler::buildArrayShape()` to parse dot-notation keys into a tree structure (`ValidationRuleNode`), then recursively build nested `TKeyedArray` unions.

- `insertIntoTree()` — splits dot-separated segments and builds a tree
- `buildUnionFromTree()` — walks the tree depth-first, producing nested shapes
- Wildcard patterns (`tags.*`, `items.*.id`) are still resolved upstream by `ValidationRuleAnalyzer::resolveRules()` — the tree builder only handles plain dot-notation nesting

Verified with 13 new type tests covering: basic nesting, mixed flat+nested, deep nesting, optional/nullable nested fields, single-field dot access, inline validate, wildcard+nested combo, parent-with-own-rule, `safe()` with nested keys, `ValidatedInput::input()` with dot keys, and optional parent groups.

## Checklist
- [x] Tests cover the change (type test in `tests/Type/`)

### CODE REVIEW

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 1]
```diff
@@ -289,29 +289,105 @@ private static function resolveFieldType(
...
+                    continue;
+                }
 
-            // Fields without a presence rule (required, present, accepted, etc.)
-            // may be absent from validated() output when not provided in the request.
-            if (!$resolvedRule->required) {
-                $fieldType = $fieldType->setPossiblyUndefined(true);
+                $fieldType = $child->rule->type;
+                if (!$child->rule->required) {
+                    $fieldType = $fieldType->setPossiblyUndefined(true);
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:367]: `setPossiblyUndefined()` does NOT mutate in-place — it returns a clone ([Union.php:305-316](https://github.com/vimeo/psalm/blob/master/src/Psalm/Type/Union.php#L305-L316)). The cached `ResolvedRule::type` is never modified. No change needed.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 2]
```diff
@@ -289,29 +289,105 @@ private static function resolveFieldType(
...
+
+            // Branch node — recurse into nested children.
+            $nested = self::buildUnionFromTree($child);
+
+            if ($nested instanceof Union) {
+                if ($child->rule !== null && !$child->rule->required) {
+                    $nested = $nested->setPossiblyUndefined(true);
+                }
+
+                $properties[$key] = $nested;
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:382]: Good catch — fixed in d04b0c8. Now uses `$rule->sometimes || !$rule->required` for both leaf and branch nodes, matching `ValidationRuleAnalyzer::resolveRules()` wildcard path.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 3]
```diff
@@ -289,29 +289,105 @@ private static function resolveFieldType(
...
+     *
+     * Wildcard patterns (tags.*, items.*.id) are already resolved by
+     * ValidationRuleAnalyzer::resolveRules() before reaching this method,
+     * so this method only handles plain dot-notation nesting.
+     *
+     * Known limitation: when all descendants of an intermediate node are optional
+     * (sometimes/no required), the parent group could be absent from validated() output,
+     * but we don't mark it as possibly_undefined. Fixing this would require checking
+     * all descendants recursively — deferred for simplicity.
+     *
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:342]: This is documented as a known limitation (lines 338-341). Computing parent optionality recursively adds significant complexity for an edge case. Tracked for follow-up.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 4]
```diff
@@ -289,29 +289,105 @@ private static function resolveFieldType(
...
+     * @psalm-mutation-free
+     */
+    private static function buildUnionFromTree(ValidationRuleNode $node): ?Union
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:345]: `setPossiblyUndefined()` returns a clone, not a mutation — see [Union.php:313](https://github.com/vimeo/psalm/blob/master/src/Psalm/Type/Union.php#L313). The input `$node` is never modified. The `@psalm-mutation-free` annotation is correct, confirmed by Psalm self-analysis passing without errors.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 5]
```diff
@@ -289,29 +289,105 @@ private static function resolveFieldType(
...
+    {
+        if ($node->children === []) {
+            return null;
+        }
+
         /** @var non-empty-array<string, Union> $properties */
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php]: Good catch — fixed. The first annotation is now `array<string, Union>`, with the `non-empty-array` assertion only after the emptiness check on line 390.

File: tests/Type/tests/ValidationTypeNarrowingTest.phpt [hunk 6]
```diff
@@ -149,5 +149,196 @@ function testInlineValidateArrayFormat(\Illuminate\Http\Request $request): void
...
+}
+
+function testParentWithOwnRuleAndChildren(ParentWithOwnRuleRequest $request): void
+{
+    // Parent 'user' has its own 'array' rule, but children provide the precise shape.
+    // The nested shape takes precedence over the parent's generic array type.
+    $_all = $request->validated();
+    /** @psalm-check-type-exact $_all = array{user: array{email: string, name: string}} */
+}
+
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply tests/Type/tests/ValidationTypeNarrowingTest.phpt:298]: Valid edge case. Currently the parent's nullable type is discarded when children provide a more precise nested shape — the result would be `array{user?: array{name?: string}}` (possibly-undefined because not required), not `array{user: array{...}|null}`. Propagating nullable from the parent rule to the nested shape would require additional logic. Tracked as follow-up alongside the parent optionality limitation.
